### ISSUE TITLE
Evaluation Details for MMAU

### ISSUE THREAD

**Issue by ChencongZJU** [QwenLM/Qwen3-Omni#63]
### Description

Thanks for your wonderful job!
I tried to reproduce the metric result of Qwen3-Omni-30B-A3B-Instruct for MMAU test-mini metrcis, but I only got accuracy of 55.40% among 1000 samples. I wonder if I set wrong inference parameters, system prompt or conversation format.
Looking forward to your reply. Best wishes!

### Reproduction

```python
import json
import torch
from typing import Dict, List, Any
from qwen_omni_utils import process_mm_info
from transformers import Qwen3OmniMoeForConditionalGeneration, Qwen2_5OmniProcessor, Qwen3OmniMoeThinkerForConditionalGeneration, Qwen3OmniMoeProcessor
import argparse
from pathlib import Path
from tqdm import tqdm

ROOT_DIR="/dllab/user/zhumuzhi/xiaomi/test-audios"


def load_model_and_processor(model_path: str) -> tuple:
    """Load Qwen3 Omni model and processor."""
    model = Qwen3OmniMoeForConditionalGeneration.from_pretrained(
        model_path,
        dtype=torch.bfloat16,
        device_map="auto",
        attn_implementation="flash_attention_2",
    )
    processor = Qwen3OmniMoeProcessor.from_pretrained(model_path)
    return model, processor


def inference(
    model,
    processor,
    audio_path: str,
    prompt: str,
    sys_prompt: str = "You are a helpful assistant."
) -> str:
    """Run inference on a single audio sample."""
    messages = [
        {"role": "system", "content": [{"type": "text", "text": sys_prompt}]},
        {
            "role": "user",
            "content": [
                {"type": "audio", "audio": audio_path},
                {"type": "text", "text": prompt},
            ],
        },
    ]
    
    text = processor.apply_chat_template(
        messages, tokenize=False, add_generation_prompt=True
    )
    
    audios, images, videos = process_mm_info(messages, use_audio_in_video=True)
    inputs = processor(
        text=text,
        audio=audios,
        images=images,
        videos=videos,
        return_tensors="pt",
        padding=True,
        use_audio_in_video=True,
    )
    inputs = inputs.to(model.device).to(model.dtype)

    output = model.generate(
        **inputs,
        use_audio_in_video=True,
        return_audio=False,
        thinker_max_new_tokens=256,
        thinker_do_sample=False,
        max_new_tokens=512,
        do_sample=False,
    )
    # import pdb; pdb.set_trace()
    text_output = processor.batch_decode(output[0], skip_special_tokens=True, clean_up_tokenization_spaces=False)
    
    # Extract only the generated part (remove the input prompt)
    generated_text = text_output[0].split("assistant\n")[-1].strip()
    print(generated_text)
    return generated_text


def format_multiple_choice_prompt(question: str, choices: List[str]) -> str:
    """Format the question and choices for the model."""
    choice_text = "\n".join([f"{chr(65+i)}. {choice}" for i, choice in enumerate(choices)])
    prompt = f"{question}\n\n{choice_text}\n\nPlease select the correct answer (A, B, C, or D):"
    return prompt


def extract_choice_from_response(response: str, choices: List[str]) -> str:
    """Extract the choice from model response."""
    response_upper = response.upper().strip()
    
    # Look for explicit choice letters (A, B, C, D)
    for i, choice_letter in enumerate(['A', 'B', 'C', 'D']):
        if choice_letter in response_upper:
            if i < len(choices):
                return choices[i]
    
    # Fallback: look for exact choice matches in response
    for choice in choices:
        if choice.lower() in response.lower():
            return choice
    
    # If no match found, return the raw response
    return response.strip()


def evaluate_mmau_subset(
    model_path: str,
    data_file: str,
    output_file: str,
    sys_prompt: str = "You are an expert audio analyst. Listen carefully to the audio and answer the question accurately."
) -> None:
    """Evaluate the model on MMAU test-mini subset."""
    
    # Load model and processor
    print("Loading model and processor...")
    model, processor = load_model_and_processor(model_path)
    
    # Load test data
    print(f"Loading test data from {data_file}...")
    with open(data_file, 'r') as f:
        test_data = json.load(f)
    
    print(f"Total samples: {len(test_data)}")
    
    # Process each sample
    results = []
    correct_predictions = 0
    total_predictions = 0
    
    for sample in tqdm(test_data, desc="Evaluating samples"):
        sample_id = sample['id']
        audio_path = sample['audio_id']
        question = sample['question']
        choices = sample['choices']
        ground_truth = sample['answer']
        
        # Format the prompt
        prompt = format_multiple_choice_prompt(question, choices)

        # audio_path=str(ROOT_DIR / audio_path)
        
        try:
            # Get model prediction
            model_response = inference(model, processor, audio_path, prompt, sys_prompt)
            model_prediction = extract_choice_from_response(model_response, choices)
            
            # Check if prediction is correct
            is_correct = model_prediction.strip().lower() == ground_truth.strip().lower()
            if is_correct:
                correct_predictions += 1
            total_predictions += 1
            
            # Create result entry
            result_sample = sample.copy()
            result_sample['model_prediction'] = model_prediction
            result_sample['model_response'] = model_response
            result_sample['is_correct'] = is_correct
            
            results.append(result_sample)
            
            print(f"Sample {sample_id}: {'✓' if is_correct else '✗'}")
            print(f"  Question: {question[:100]}...")
            print(f"  Ground truth: {ground_truth}")
            print(f"  Model prediction: {model_prediction}")
            print(f"  Model response: {model_response[:100]}...")
            print()
            
        except Exception as e:
            print(f"Error processing sample {sample_id}: {e}")
            # Add failed sample with error info
            result_sample = sample.copy()
            result_sample['model_prediction'] = "ERROR"
            result_sample['model_response'] = str(e)
            result_sample['is_correct'] = False
            results.append(result_sample)
    
    # Calculate accuracy
    accuracy = correct_predictions / total_predictions if total_predictions > 0 else 0.0
    
    # Add evaluation metadata
    evaluation_results = {
        "metadata": {
            "model_path": model_path,
            "total_samples": len(test_data),
            "processed_samples": total_predictions,
            "correct_predictions": correct_predictions,
            "accuracy": accuracy,
            "system_prompt": sys_prompt
        },
        "results": results
    }
    
    # Save results
    print(f"Saving results to {output_file}...")
    with open(output_file, 'w') as f:
        json.dump(evaluation_results, f, indent=2, ensure_ascii=False)
    
    print(f"Evaluation completed!")
    print(f"Accuracy: {accuracy:.4f} ({correct_predictions}/{total_predictions})")
    print(f"Results saved to: {output_file}")


def main():
    parser = argparse.ArgumentParser(description="Evaluate Qwen2.5 Omni on MMAU test-mini subset")
    parser.add_argument(
        "--model_path", 
        type=str, 
        default="/dllab/user/zhumuzhi/weights/Qwen3-Omni-30B-A3B-Instruct/",
    )
    parser.add_argument(
        "--data_file", 
        type=str, 
        default="mmau-test-mini.json",
        help="Path to the MMAU test-mini JSON file"
    )
    parser.add_argument(
        "--output_file", 
        type=str, 
        default="mmau_evaluation_results_av_reorder_fix_mini.json",
        help="Path to save the evaluation results"
    )
    parser.add_argument(
        "--sys_prompt",
        type=str,
        default="You are an expert audio analyst. Listen carefully to the audio and answer the question accurately.",
        help="System prompt for the model"
    )
    
    args = parser.parse_args()
    
    evaluate_mmau_subset(
        model_path=args.model_path,
        data_file=args.data_file,
        output_file=args.output_file,
        sys_prompt=args.sys_prompt
    )


if __name__ == "__main__":
    main()
``` 

### Logs

```shell

```

### Environment Information

H20

### Known Issue

- [ ] The issue hasn't been already addressed in Documentation, Issues, and Discussions.

**Comment by BakerBunker** [comment]
During evaluation, we did not use a system prompt, and we used vLLM for inference, which may introduce minor precision differences compared to Hugging Face's implementation.

Answer extraction code:
```python
def string_match(answer, prediction, choices):
    # Function to normalize and tokenize text
    def tokenize(text):
        # Convert to lowercase and find all word tokens
        return set(re.findall(r"\b\w+\b", text.lower()))

    # Tokenize prediction and answer
    prediction_tokens = tokenize(prediction)
    answer_tokens = tokenize(answer)

    if not prediction_tokens:
        return False

    # Tokenize incorrect choices and exclude tokens present in the answer
    incorrect_tokens = set()
    for choice in choices:
        choice_tokens = tokenize(choice)
        if choice_tokens != answer_tokens:
            incorrect_tokens.update(choice_tokens - answer_tokens)

    # Condition 1: All tokens of the answer are in the prediction
    cond1 = answer_tokens.issubset(prediction_tokens)

    # Condition 2: Prediction does not contain any tokens from incorrect choices (excluding shared words)
    cond2 = prediction_tokens.isdisjoint(incorrect_tokens)

    return cond1 and cond2
```

Sample input:
```json
[
        [
            {
                "type": "audio",
                "audio": "mmau/test-mini-audios/3fe64f3d-282c-4bc8-a753-68f8f6c35652.wav"
            },
            {
                "type": "text",
                "text": "Based on the given audio, identify the source of the speaking voice. Select one option from the provided choices.\n['Man', 'Woman', 'Child', 'Robot']"
            }
        ]
    ]
```

**Comment by ChencongZJU** [comment]
> During evaluation, we did not use a system prompt, and we used vLLM for inference, which may introduce minor precision differences compared to Hugging Face's implementation.
> 
> Answer extraction code:
> 
> def string_match(answer, prediction, choices):
>     # Function to normalize and tokenize text
>     def tokenize(text):
>         # Convert to lowercase and find all word tokens
>         return set(re.findall(r"\b\w+\b", text.lower()))
> 
>     # Tokenize prediction and answer
>     prediction_tokens = tokenize(prediction)
>     answer_tokens = tokenize(answer)
> 
>     if not prediction_tokens:
>         return False
> 
>     # Tokenize incorrect choices and exclude tokens present in the answer
>     incorrect_tokens = set()
>     for choice in choices:
>         choice_tokens = tokenize(choice)
>         if choice_tokens != answer_tokens:
>             incorrect_tokens.update(choice_tokens - answer_tokens)
> 
>     # Condition 1: All tokens of the answer are in the prediction
>     cond1 = answer_tokens.issubset(prediction_tokens)
> 
>     # Condition 2: Prediction does not contain any tokens from incorrect choices (excluding shared words)
>     cond2 = prediction_tokens.isdisjoint(incorrect_tokens)
> 
>     return cond1 and cond2
> Sample input:
> 
> [
>         [
>             {
>                 "type": "audio",
>                 "audio": "mmau/test-mini-audios/3fe64f3d-282c-4bc8-a753-68f8f6c35652.wav"
>             },
>             {
>                 "type": "text",
>                 "text": "Based on the given audio, identify the source of the speaking voice. Select one option from the provided choices.\n['Man', 'Woman', 'Child', 'Robot']"
>             }
>         ]
>     ]

Thanks for your reply! I will try to re-evaluate using vLLM right now. I have another question about detailed vLLM sampling parameters. As in https://github.com/QwenLM/Qwen3-Omni#evaluation, it says Instruct models use greedy decoding during generation without sampling.
Looking forward to your reply. Best wishes!

**Comment by BakerBunker** [comment]
Yes, we indeed use greedy decoding during instruction model evaluation.

**Comment by ChencongZJU** [comment]
> Yes, we indeed use greedy decoding during instruction model evaluation.

Got it. I will set sampling parameters to avoid sampling randomness.
Thanks again~

**Comment by ChencongZJU** [comment]
By substitute the question and answer match function. We reproduce the metric of Qwen3-Omni-30B-A3B-Instruct on MMAU Benchmark with precision of 78.8%. We use transformers and code is following:
Thanks for your help!


```python
import json
import torch
from typing import Dict, List, Any
from qwen_omni_utils import process_mm_info
from transformers import Qwen3OmniMoeForConditionalGeneration, Qwen3OmniMoeProcessor
import argparse
from pathlib import Path
from tqdm import tqdm
import re

ROOT_DIR="/dllab/user/zhumuzhi/xiaomi/test-audios"


def load_model_and_processor(model_path: str) -> tuple:
    """Load Qwen3 Omni model and processor."""
    model = Qwen3OmniMoeForConditionalGeneration.from_pretrained(
        model_path,
        dtype=torch.bfloat16,
        device_map="auto",
        attn_implementation="flash_attention_2",
    )
    processor = Qwen3OmniMoeProcessor.from_pretrained(model_path)
    return model, processor


def inference(
    model,
    processor,
    audio_path: str,
    prompt: str,
) -> str:
    """Run inference on a single audio sample."""
    messages = [
        {
            "role": "user",
            "content": [
                {"type": "audio", "audio": audio_path},
                {"type": "text", "text": prompt},
            ],
        },
    ]
    
    text = processor.apply_chat_template(
        messages, tokenize=False, add_generation_prompt=True
    )
    
    audios, images, videos = process_mm_info(messages, use_audio_in_video=True)
    inputs = processor(
        text=text,
        audio=audios,
        images=images,
        videos=videos,
        return_tensors="pt",
        padding=True,
        use_audio_in_video=True,
    )
    inputs = inputs.to(model.device).to(model.dtype)

    output = model.generate(
        **inputs,
        use_audio_in_video=True,
        return_audio=False,
        thinker_max_new_tokens=256,
        thinker_do_sample=False,
        max_new_tokens=512,
        do_sample=False,
    )
    # import pdb; pdb.set_trace()
    text_output = processor.batch_decode(output[0], skip_special_tokens=True, clean_up_tokenization_spaces=False)
    
    # Extract only the generated part (remove the input prompt)
    generated_text = text_output[0].split("assistant\n")[-1].strip()
    print(generated_text)
    return generated_text


def format_multiple_choice_prompt(question: str, choices: List[str]) -> str:
    """Format the question and choices for the model."""
    prompt = f"{question}Select one option from the provided choices.{choices}"
    return prompt


def string_match(answer, prediction, choices):
    # Function to normalize and tokenize text
    def tokenize(text):
        # Convert to lowercase and find all word tokens
        return set(re.findall(r"\b\w+\b", text.lower()))

    # Tokenize prediction and answer
    prediction_tokens = tokenize(prediction)
    answer_tokens = tokenize(answer)

    if not prediction_tokens:
        return False

    # Tokenize incorrect choices and exclude tokens present in the answer
    incorrect_tokens = set()
    for choice in choices:
        choice_tokens = tokenize(choice)
        if choice_tokens != answer_tokens:
            incorrect_tokens.update(choice_tokens - answer_tokens)

    # Condition 1: All tokens of the answer are in the prediction
    cond1 = answer_tokens.issubset(prediction_tokens)

    # Condition 2: Prediction does not contain any tokens from incorrect choices (excluding shared words)
    cond2 = prediction_tokens.isdisjoint(incorrect_tokens)

    return cond1 and cond2

def evaluate_mmau_subset(
    model_path: str,
    data_file: str,
    output_file: str,
) -> None:
    """Evaluate the model on MMAU test-mini subset."""
    
    # Load model and processor
    print("Loading model and processor...")
    model, processor = load_model_and_processor(model_path)
    
    # Load test data
    print(f"Loading test data from {data_file}...")
    with open(data_file, 'r') as f:
        test_data = json.load(f)
    
    print(f"Total samples: {len(test_data)}")
    
    # Process each sample
    results = []
    correct_predictions = 0
    total_predictions = 0
    
    for sample in tqdm(test_data, desc="Evaluating samples"):
        sample_id = sample['id']
        audio_path = sample['audio_id']
        question = sample['question']
        choices = sample['choices']
        ground_truth = sample['answer']
        
        # Format the prompt
        prompt = format_multiple_choice_prompt(question, choices)

        # audio_path=str(ROOT_DIR / audio_path)
        
        try:
            # Get model prediction
            model_response = inference(model, processor, audio_path, prompt)
            model_prediction = model_response
            
            # Check if prediction is correct
            is_correct = string_match(model_response, ground_truth, choices)
            if is_correct:
                correct_predictions += 1
            total_predictions += 1
            
            # Create result entry
            result_sample = sample.copy()
            result_sample['model_prediction'] = model_prediction
            result_sample['model_response'] = model_response
            result_sample['is_correct'] = is_correct
            
            results.append(result_sample)
            
            print(f"Sample {sample_id}: {'✓' if is_correct else '✗'}")
            print(f"  Question: {question[:100]}...")
            print(f"  Ground truth: {ground_truth}")
            print(f"  Model prediction: {model_prediction}")
            print(f"  Model response: {model_response[:100]}...")
            print()
            
        except Exception as e:
            print(f"Error processing sample {sample_id}: {e}")
            # Add failed sample with error info
            result_sample = sample.copy()
            result_sample['model_prediction'] = "ERROR"
            result_sample['model_response'] = str(e)
            result_sample['is_correct'] = False
            results.append(result_sample)
    
    # Calculate accuracy
    accuracy = correct_predictions / total_predictions if total_predictions > 0 else 0.0
    
    # Add evaluation metadata
    evaluation_results = {
        "metadata": {
            "model_path": model_path,
            "total_samples": len(test_data),
            "processed_samples": total_predictions,
            "correct_predictions": correct_predictions,
            "accuracy": accuracy,
        },
        "results": results
    }
    
    # Save results
    print(f"Saving results to {output_file}...")
    with open(output_file, 'w') as f:
        json.dump(evaluation_results, f, indent=2, ensure_ascii=False)
    
    print(f"Evaluation completed!")
    print(f"Accuracy: {accuracy:.4f} ({correct_predictions}/{total_predictions})")
    print(f"Results saved to: {output_file}")


def main():
    parser = argparse.ArgumentParser(description="Evaluate Qwen2.5 Omni on MMAU test-mini subset")
    parser.add_argument(
        "--model_path", 
        type=str, 
        default="/dllab/user/zhumuzhi/weights/Qwen3-Omni-30B-A3B-Instruct/",
    )
    parser.add_argument(
        "--data_file", 
        type=str, 
        default="mmau-test-mini.json",
        help="Path to the MMAU test-mini JSON file"
    )
    parser.add_argument(
        "--output_file", 
        type=str, 
        default="mmau_evaluation_results_av_reorder_fix_mini.json",
        help="Path to save the evaluation results"
    )
    
    args = parser.parse_args()
    
    evaluate_mmau_subset(
        model_path=args.model_path,
        data_file=args.data_file,
        output_file=args.output_file,
    )


if __name__ == "__main__":
    main()
```
